tests/unit/state_manager/test_state_manager.py

```diff
@@ -172,3 +176,79 @@ def test_set_step_updated(self):
 
         rewrapped_stw = sdb.get(part_name="foo", step=Step.PULL)
         assert rewrapped_stw.serial == 2
+
+
+@pytest.mark.usefixtures("new_dir")
+class TestStateManager:
+    """Verify if the State Manager is correctly tracking state changes."""
+
+    def test_has_step_run(self):
+        info = ProjectInfo()
+        p1 = Part("p1", {})
+        p2 = Part("p2", {})
+        p3 = Part("p3", {})
+
+        s3 = states.StageState()
+        s3.write(Path("parts/p3/state/stage"))
+
+        # only p3:stage has a state file when the stage manager is created
+        sm = StateManager(project_info=info, part_list=[p1, p2, p3])
+
+        for part in [p1, p2, p3]:
+            for step in list(Step):
+                ran = sm.has_step_run(part, step)
+                assert ran == (part == p3 and step == Step.STAGE)
```

Comment:
if only step 3 has run, then p1 and p2 should have an assertion that they have not

Reply:
The nested for loops iterate over all steps for all parts, and the assertion tests them all against False, except for the step that statisfies the condition (part p3 and step stage), which is asserted True.